types/src/ledger_info.rs

```diff
@@ -237,70 +255,146 @@ impl LedgerInfoWithV0 {
     }
 
-    pub fn add_signature(&mut self, validator: AccountAddress, signature: bls12381::Signature) {
-        self.signatures.entry(validator).or_insert(signature);
+    pub fn get_voters(&self, validator_addresses: &Vec<AccountAddress>) -> Vec<AccountAddress> {
+        self.signatures.get_voter_addresses(validator_addresses)
     }
 
-    pub fn remove_signature(&mut self, validator: AccountAddress) {
-        self.signatures.remove(&validator);
+    pub fn get_num_voters(&self) -> usize {
+        self.signatures.get_num_voters()
     }
 
-    pub fn signatures(&self) -> &BTreeMap<AccountAddress, bls12381::Signature> {
-        &self.signatures
+    pub fn get_voters_bitmap(&self) -> &Vec<bool> {
+        self.signatures.get_voters_bitmap()
     }
 
     pub fn verify_signatures(
         &self,
         validator: &ValidatorVerifier,
     ) -> ::std::result::Result<(), VerifyError> {
-        validator.batch_verify_aggregated_signatures(self.ledger_info(), self.signatures())
+        validator.verify_aggregated_signatures(self.ledger_info(), &self.signatures)
     }
 
     pub fn check_voting_power(
         &self,
         validator: &ValidatorVerifier,
     ) -> ::std::result::Result<(), VerifyError> {
-        validator.check_voting_power(self.signatures.keys())
+        validator.check_voting_power(
+            self.get_voters(&validator.get_ordered_account_addresses_iter().collect_vec())
+                .iter(),
+        )
+    }
+    pub fn signatures(&self) -> &AggregatedSignature {
+        &self.signatures
+    }
+}
+
+/// Contains the ledger info and partially aggregated signature from a set of validators, this data
+/// is only used during the aggregating the votes from different validators and is not persisted in
+/// DB.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct LedgerInfoWithPartialSignatures {
+    ledger_info: LedgerInfo,
+    partial_sigs: PartialSignatures,
+}
+
+impl Display for LedgerInfoWithPartialSignatures {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        write!(f, "{}", self.ledger_info)
+    }
+}
+
+impl LedgerInfoWithPartialSignatures {
+    pub fn new(ledger_info: LedgerInfo, signatures: PartialSignatures) -> Self {
+        Self {
+            ledger_info,
+            partial_sigs: signatures,
+        }
+    }
+
+    pub fn commit_info(&self) -> &BlockInfo {
+        self.ledger_info.commit_info()
+    }
+
+    pub fn remove_signature(&mut self, validator: AccountAddress) {
+        self.partial_sigs.remove_signature(validator);
+    }
+
+    pub fn add_signature(&mut self, validator: AccountAddress, signature: bls12381::Signature) {
+        self.partial_sigs.add_signature(validator, signature);
+    }
+
+    pub fn signatures(&self) -> &HashMap<AccountAddress, bls12381::Signature> {
+        self.partial_sigs.signatures()
+    }
+
+    pub fn aggregate_signatures<T: CryptoHash + Serialize>(
+        &self,
+        verifier: &ValidatorVerifier,
+        message: &T,
```

Comment:
why this takes a message? it should pass self.ledger_info directly?

Reply:
A bunch of unit tests are calling this function without ledger_info.

Reply:
can you give some examples? this struct carries a ledger info, aggregate without ledger info should use the verifier function instead

Reply:
Actually there is just one test IIUC `test_block_metadata_bitmaps`. If you have strong opinion, I can take a stab at changing that test.

Reply:
I don't know what that test needs to sign a different message, but let's change it

Reply:
Changed